### PR TITLE
PVF: Don't dispute on missing artifact

### PR DESCRIPTION
**Pull request by mrcnski**
# PULL REQUEST

## Overview

A dispute should never be raised if the local cache doesn't provide a certain artifact. You can not dispute based on this reason, as it is a local hardware issue and not related to the candidate to check.

## Design

Currently we assume that if we prepared an artifact, it remains there on-disk until we prune it, i.e. we never check again if it's still there.

We can change it so that instead of artifact-not-found triggering a dispute, we retry once (like we do for `AmbiguousWorkerDeath`, except we don't dispute if it still doesn't work). And when enqueuing an execute job, we check for the artifact on-disk, and start preparation if not found.

## Changes

- [x] Integration test (should fail without the following changes)
- [x] Check if artifact exists when executing, prepare if not
- [x] Return an internal error when file is missing
- [x] Retry once on internal errors
- [x] Document design (update impl guide)
- [x] Also snuck in a refactor of `handle_execute_pvf`...

## Related issues

Closes https://github.com/paritytech/polkadot/issues/6959
Pre-requisite for https://github.com/paritytech/polkadot/issues/6940

### CODE REVIEW

File: node/core/pvf/src/execute/worker.rs [hunk 1]
```diff
@@ -359,7 +366,13 @@ fn validate_using_artifact(
 		//         [`executor_intf::prepare`].
 		executor.execute(artifact_path.as_ref(), params)
 	} {
-		Err(err) => return Response::format_invalid("execute", &err),
+		Err(err) =>
+			return if err.contains("failed to open file: No such file or directory") {
+				// Raise an internal error if the file is missing.
+				Response::format_internal("missing file", &err)
+			} else {
+				Response::format_invalid("execute", &err)
```
Comment (mrcnski) [inline node/core/pvf/src/execute/worker.rs:374]: Before we always raised `Invalid` which would dispute. Now we raise `Internal` which does not dispute.

File: node/core/candidate-validation/src/lib.rs [hunk 2]
```diff
@@ -691,38 +691,54 @@ trait ValidationBackend {
...
-		// If we get an AmbiguousWorkerDeath error, retry once after a brief delay, on the
-		// assumption that the conditions that caused this error may have been transient. Note that
-		// this error is only a result of execution itself and not of preparation.
-		if let Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::AmbiguousWorkerDeath)) =
-			validation_result
-		{
-			// Wait a brief delay before retrying.
-			futures_timer::Delay::new(PVF_EXECUTION_RETRY_DELAY).await;
+		// Allow one retry for each kind of error.
+		let mut num_internal_retries_left = 1;
```
Comment (mrcnski) [inline node/core/candidate-validation/src/lib.rs:712]: Could make this higher, since this kind of error is probably the most likely to be transient.

File: node/core/pvf/src/execute/worker.rs (hunk 1 above)
Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/execute/worker.rs]: Have you checked the `No such file or directory` part with a different system locale? I'm not sure how it's generated, if it relies on libc's `strerror(errno)` it could be different if the system locale differs, on my system it reads `File o directory non esistente` (because of Italian locale), but I didn't check if that's the case for this exact error message.

File: node/core/pvf/src/execute/worker.rs [hunk 3]
```diff
@@ -359,7 +366,13 @@ fn validate_using_artifact(
 		//         [`executor_intf::prepare`].
 		executor.execute(artifact_path.as_ref(), params)
 	} {
-		Err(err) => return Response::format_invalid("execute", &err),
+		Err(err) =>
+			return if err.contains("failed to open file: No such file or directory") {
```
Comment (eskimor) [inline node/core/pvf/src/execute/worker.rs:370]: This really requires a refactor changing the error type to something sensible, like an `enum`.  Matching on a string is way too error prone. Something changes the message, localization, ....

Comment (slumber) [reply node/core/pvf/src/execute/worker.rs:370]: Very likely to come from substrate, it's full of string errors. I agree that matching against strings is no-go, but otherwise we'd have to halt the pr.

Comment (s0me0ne-unkn0wn) [reply node/core/pvf/src/execute/worker.rs:370]: > This really requires a refactor changing the error type to something sensible, like an `enum`

Vote up, I've already raised that concern somewhere... Many errors coming from Substrate are not sensible at all. Also agree that string matching makes no good.

@mrcnski a (probably stupid) idea: until we have an `enum` error from Substrate, would it be better not to rely on its string errors but to check for the file existence ourselves? Should be simple enough. Of course, it introduces a race condition, but still better than parsing strings. Also, nobody guarantees that the file persists between the moments when it is open and when it is read, so that kind of race condition already exists anyway.

Comment (mrcnski) [reply node/core/pvf/src/execute/worker.rs:370]: Yeah, it comes from Substrate. Didn't think about localization. 😬 I considered just treating `RuntimeConstruction` itself as an internal error, but seems it's also used for some case where wasm runs out of memory, which would be a problem with the PVF itself. [link](https://github.com/paritytech/substrate/blob/master/client/executor/src/integration_tests/mod.rs#L824)

Checking for the file existence seems sensible to me...

Comment (eskimor) [reply node/core/pvf/src/execute/worker.rs:370]: Problems with the PVF itself we agreed are also no reason to raise a dispute, since we have pre-checking enabled. Basically any error that is independent of the candidate at hand should not be cause for a dispute.

Comment (eskimor) [reply node/core/pvf/src/execute/worker.rs:370]: Let's in any case create a ticket for fixing those string errors - or at least the one in question right now.

Comment (mrcnski) [reply node/core/pvf/src/execute/worker.rs:370]: Looks like `"output exceeds bounds of wasm memory"` should not be an issue with runtime construction, but is actually indicative of a malicious PVF. And I think this would get past pre-checking, since that just compiles and doesn't execute.

https://github.com/paritytech/substrate/blob/master/client/executor/wasmtime/src/runtime.rs#L772-L780

```rs
// Do a length check before allocating. The returned output should not be bigger than the
// available WASM memory. Otherwise, a malicious parachain can trigger a large allocation,
// potentially causing memory exhaustion.
//
// Get the size of the WASM memory in bytes.
let memory_size = ctx.as_context().data().memory().data_size(ctx);
if checked_range(output_ptr as usize, output_len as usize, memory_size).is_none() {
    Err(WasmError::Other("output exceeds bounds of wasm memory".into()))?
}
```

(I used `WasmError::Other` here to match the other errors in the file, without realizing it gets converted to `RuntimeConstruction`. 🤷‍♂️)

Anyway, basically, this one `"output exceeds bounds of wasm memory"` case is deterministic and we should definitely vote against. If we gave it a new separate enum in Substrate, then we could treat the existing `RuntimeConstruction` as an internal error.

Comment (mrcnski) [reply node/core/pvf/src/execute/worker.rs:370]: Raised a small fix for the output-bounds case [here](https://github.com/paritytech/substrate/pull/13841), but I'm still not confident that `RuntimeConstruction` is always a transient error and don't think we should use it. Unless we treat it as a new "possibly transient" case, meaning we retry and dispute if it happens again. 🤷‍♂️

For the file-not-found case, there is not a clear way to fix the error story on the Substrate side. Just having another check here for file existence should be enough.

Comment (eskimor) [reply node/core/pvf/src/execute/worker.rs]: I am fine with the check, please reference the substrate issue though in a comment. This way, readers understand why we did it this way and we can reevaluate once the issue is fixed.

File: node/core/pvf/src/execute/worker.rs [hunk 4]
```diff
@@ -353,6 +360,11 @@ fn validate_using_artifact(
 	executor: Arc<Executor>,
 	cpu_time_start: ProcessTime,
 ) -> Response {
+	let file_metadata = std::fs::metadata(artifact_path);
+	if let Err(err) = file_metadata {
+		return Response::format_internal("execute: could not find or open file", &err.to_string())
```
Comment (eskimor) [inline node/core/pvf/src/execute/worker.rs:365]: What if the file is there, but corrupted? Do we handle that case correctly already?

Comment (mrcnski) [reply node/core/pvf/src/execute/worker.rs:365]: Very good question. I believe right now it would result in a panic. There's an open issue about it at https://github.com/paritytech/polkadot/issues/6972

Comment (eskimor) [reply node/core/pvf/src/execute/worker.rs:365]: Nice! Thank you for staying on top of things! :1st_place_medal:

**Comment by mrcnski** [comment]
Ready for another review. :) (Don't think we can do much about the Substrate error for now.)

File: node/core/candidate-validation/src/lib.rs [hunk 5]
```diff
@@ -691,38 +691,54 @@ trait ValidationBackend {
```
Comment (slumber) [inline node/core/candidate-validation/src/lib.rs:691]: Updating candidate-validation tests would not harm

**Comment by mrcnski** [comment]
bot merge

**Comment by mrcnski** [comment]
bot rebase

**Comment by mrcnski** [comment]
bot merge
